base/rational.jl

```diff
@@ -129,12 +129,344 @@ Bool(x::Rational) = x==0 ? false : x==1 ? true :
 (::Type{T})(x::Rational) where {T<:Integer} = (isinteger(x) ? convert(T, x.num)::T :
     throw(InexactError(nameof(T), T, x)))
 
-AbstractFloat(x::Rational) = (float(x.num)/float(x.den))::AbstractFloat
-function (::Type{T})(x::Rational{S}) where T<:AbstractFloat where S
-    P = promote_type(T,S)
-    convert(T, convert(P,x.num)/convert(P,x.den))::T
+bit_width(n) = ndigits(n, base = UInt8(2), pad = false)
+
+function divrem_pow2(num::Integer, n::Integer)
+    quot = num >> n
+    rema = num - (quot << n)
+    (quot, rema)
 end
 
+module RationalToFloatRoundingMode
+    struct R
+        r::Int
+        function R(r::Int)
+            (1 ≤ r ≤ 4) || error("unknown rounding mode")
```

Comment:
If this construction is here to stay:

```suggestion
            (1 ≤ r ≤ 4) || throw(ArgumentError("Invalid rounding mode"))
```

Though maybe an `@enum` would be better suited.

Reply:
I think the benefit of a module over an `@enum` is mostly that we get the scoping.

Reply:
The `@enum` can live in a custom module just fine - though again, I don't follow the design advantage over using the existing `RoundingMode` and branching appropriately. Branches checking types will be optimized away in type stable code; they don't end up in the final assembly.